### ISSUE TITLE
More ways to make CI faster?

### ISSUE THREAD

**Issue by hossain-khan** [hossain-khan/android-timeline#16]
To speed up your Gradle builds in GitHub Actions, there are a few strategies you can implement to reduce the time it takes to start the Gradle Daemon and overall build times:

### 1. **Use the Gradle Cache**
You're already using `actions/cache@v4` for caching. However, you need to ensure that Gradle-specific caches are being stored and restored correctly between builds to avoid rebuilding everything from scratch every time.

Add a cache step for both the Gradle User Home and `.gradle` directories:

```yaml
jobs:
  android-build:
    runs-on: ubuntu-latest

    strategy:
      matrix:
        java-version: [ '22', '23' ]
        distribution: [ 'temurin' ]

    steps:
      - uses: actions/checkout@v4

      - name: Cache Gradle wrapper and dependencies
        uses: actions/cache@v4
        with:
          path: |
            ~/.gradle/caches
            ~/.gradle/wrapper
          key: gradle-${{ matrix.java-version }}-${{ runner.os }}-${{ hashFiles('**/*.gradle*', '**/gradle-wrapper.properties') }}
          restore-keys: |
            gradle-${{ matrix.java-version }}-${{ runner.os }}

      - name: set up JDK
        uses: actions/setup-java@v4
        with:
          java-version: ${{ matrix.java-version }}
          distribution: ${{ matrix.distribution }}

      - name: Setup Gradle
        uses: gradle/actions/setup-gradle@v4

      - name: Run Linter
        run: ./gradlew lintKotlin --daemon

      - name: Run test
        run: ./gradlew testDebugUnitTest --daemon

      - name: Build with Gradle
        run: ./gradlew assembleDebug --daemon
```

This caches Gradle files between runs to speed up dependency resolution and avoids downloading dependencies every time.

### 2. **Use `--daemon` Option**
You can explicitly enable the Gradle Daemon by adding the `--daemon` option to your Gradle commands, as shown above. This ensures that the daemon starts earlier in the build process and will be reused for subsequent tasks.

### 3. **Enable Gradle Build Caching**
You can enable Gradle's build caching feature to cache build outputs between different jobs and builds:

Add this to your `gradle.properties` file in your project:
```properties
org.gradle.caching=true
```

### 4. **Use `--parallel` Option**
If your project has multiple independent modules, you can enable parallel builds by adding the `--parallel` flag to your build commands:

```yaml
      - name: Build with Gradle
        run: ./gradlew assembleDebug --parallel --daemon
```

This will allow Gradle to run multiple tasks simultaneously, which can significantly reduce build time.

By implementing these strategies, you'll likely see faster build times in your GitHub Actions workflow.

**Comment by hossain-khan** [comment]
To improve the build speed in your GitHub Actions workflow, you can follow these recommendations:

1. **Enable the Gradle Daemon**: Ensure that the Gradle Daemon is running to speed up subsequent builds.
2. **Use Build Caching**: Enable the Gradle build cache to reuse outputs from previous builds.
3. **Optimize Dependency Resolution**: Make sure dependencies are cached effectively.

Update your `.github/workflows/android.yml` file as follows:

```yaml
name: Android CI

on:
  push:
    branches: [ "main" ]
  pull_request:
    branches: [ "main" ]

jobs:
  android-build:
    runs-on: ubuntu-latest

    strategy:
      matrix:
        java-version: [ '22', '23' ]
        distribution: [ 'temurin' ]

    steps:
      - uses: actions/checkout@v4
      - name: set up JDK
        uses: actions/setup-java@v4
        with:
          java-version: ${{ matrix.java-version }}
          distribution: ${{ matrix.distribution }}
          cache: gradle

      - name: Grant execute permission for gradlew
        run: chmod +x gradlew

      # Enable the Gradle Daemon and build cache
      - name: Configure Gradle
        run: echo "org.gradle.daemon=true\norg.gradle.caching=true" >> ~/.gradle/gradle.properties

      - name: Run Linter
        run: ./gradlew lintKotlin

      - name: Run test
        run: ./gradlew testDebugUnitTest

      - name: Build with Gradle
        run: ./gradlew assembleDebug

      - name: Cache Gradle Dependencies
        uses: actions/cache@v4
        with:
          path: |
            ~/.gradle/caches
            ~/.gradle/wrapper
          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*', '**/gradle-wrapper.properties') }}
          restore-keys: |
            ${{ runner.os }}-gradle-
```

These changes will enable the Gradle Daemon and build caching, which should help speed up your builds.

**Comment by hossain-khan** [comment]
Good hint! for now I will park it.

**Comment by hossain-khan** [comment]
Fixed in #17 

Related to #9
